### PR TITLE
[WIP] YOLO working on tinygrad!

### PR DESCRIPTION
**Pull request by skoshx**
This pull request adds yolo v3 model to tinygrad. Run it with `PYTHONPATH=. python examples/yolov3.py`

Here's what still needs to be done:

- [x] Optimizing the YOLO layer, since there is a lot of unoptimized memory copying from GPU->CPU. (right now because of it yolo runs faster on CPU than GPU)
- [x] Fixing bug with `.cpu()` function call taking up to 5 seconds, smaller tensors take longer? Makes no sense... Solving this would mean good inference times on GPU...
- [ ] Implementing strided max_pool2d & MaxPool2d layer, to add support for YOLO v4 and other YOLO models (tiny yolo for instance)
- [ ] Adding some changes to the YOLO layer described in YOLOv4 paper.
- [ ] Fix memory leak, no idea where it's leaking...
- [x] Finishing off the `process_results` function, to produce output images with bounding boxes.

### CODE REVIEW

File: tinygrad/nn.py [hunk 1]
```diff
@@ -18,6 +18,7 @@ def __call__(self, x):
     if self.track_running_stats:
       self.running_mean = (1 - self.momentum) * self.running_mean + self.momentum * batch_mean
       self.running_var = (1 - self.momentum) * self.running_var + self.momentum * batch_var
+      if self.num_batches_tracked is None: self.num_batches_tracked = Tensor.zeros(1, requires_grad=False)
```
Comment (geohot) [inline tinygrad/nn.py:21]: How could this ever happen?

**Comment by geohot** [comment]
Nice job! You can delete YoloV5 also.

File: tinygrad/nn.py (hunk 1 above)
Comment (skoshx) [reply tinygrad/nn.py:21]: Good question, I got baffled by this too... I spent a good while trying to find where `num_batches_tracked` could become None to no avail, so I added this line as a quick workaround... I still have to investigate this issue...

File: extra/yolov3.py [hunk 2]
```diff
@@ -0,0 +1,626 @@
...
+  grid_size = inp_dim // stride
+  bbox_attrs = 5 + num_classes
+  num_anchors = len(anchors)
+  
+  prediction = prediction.reshape(shape=(batch_size, bbox_attrs*num_anchors, grid_size*grid_size))
+  # Original PyTorch: transpose(1, 2) -> For some reason numpy.transpose order has to be reversed?
+  prediction = prediction.transpose(order=(0, 2, 1))
+  prediction = prediction.reshape(shape=(batch_size, grid_size*grid_size*num_anchors, bbox_attrs))
+  
+  prediction_cpu = prediction.cpu().data
```
Comment (skoshx) [inline extra/yolov3.py:279]: @geohot Any idea why this line is taking ~1s to run? Im close to getting decent results on GPU, this is the only line left taking a long time… 

When i run the model with bigger input size `width=608` the line takes approximately one second to run, however, when I run it with `width=416` (smaller input size, .cpu() should return faster) the line takes 5 seconds!! There has to be something underlying here, to point out a reason why a larger tensor takes less time to put on the CPU...

Once this is figured out, we can actually achieve some good results on webcam using GPU!

**Comment by geohot** [comment]
.cpu() is slow because that is what actually calls clFinish. The GPU calls don't block.

**Comment by skoshx** [comment]
@geohot I see. I started to write the YOLO layer using GPU ops only, but the problem is lines like these:
```python
prediction[:,:,0] = prediction[:,:,0].sigmoid()
```

The Tensor class nor the GPUBuffer class support item assignment, so I'm not sure how I should fix lines like this... I'm working on adding item assignment support, if you have any ideas or direction regarding this, I would appreciate it... Am struggling to wrap my head around these things.

**Comment by geohot** [comment]
Merged, want to add a shill to the readme? I see the jpg merged.

**Comment by skoshx** [comment]
Do you mean like a preview of the result image? If so, I do think it would sit well next to for instance the GAN image!

**Pull request merged**